feat: add cli guides (#82)
sapphiredev/website: docs/Guide/CLI/custom-templates.md

---
sidebar_position: 3
title: Create custom component templates
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem';

## Enable custom component templates

<Tabs groupId="language-choice">
<TabItem value="json" label="JSON" default>

```json {11-14}
{
  "$schema": "https://raw.githubusercontent.com/sapphiredev/cli/main/templates/schemas/.sapphirerc.scheme.json",
  "projectLanguage": "ts",
  "locations": {
    "base": "src",
    "arguments": "arguments",
    "commands": "commands",
    "listeners": "listeners",
    "preconditions": "preconditions"
  },
  "customFileTemplates": {
    "enabled": true,
    "location": "templates"
  }
}
```

</TabItem>

<TabItem value="yaml" label="YAML">

```yaml {8-10}
projectLanguage: ts
locations:
  base: src
  arguments: arguments
  commands: commands
  listeners: listeners
  preconditions: preconditions
customFileTemplates:
  enabled: true
  location: 'templates'
```

</TabItem>
</Tabs>

:::note

> If you are using json, you will get intellisense, because of our $schema

Reviewer: ```suggestion
If you are using JSON, you will get intellisense because of our $schema
```

:::

- **customFileTemplates.enabled**: Enables/disables custom component templates
- **customFileTemplates.location**: The location of your templates

## Create a template

- Create a file with a name like this in your custom template directory: `<templateName>.<language>.sapphire` (e.g
  `command.ts.sapphire`). If you make its name same as one of the default templates, your template will override the
  default one.
- Template's have 2 parts, config and the template, separated with `---`.
- We first need to type the config:

```json
{
  "category": "commands"
}

> ```

Reviewer: YAML is not supported for template configuration

> ```

Author: remove this suggestion, nyte

> ```

Reviewer 2: We  resolve suggestions, not remove them.

`category` is the category of that template, CLI uses it to know where to create the component by finding that
category's location from the `locations` field in your `.sapphirerc` file. You can create your own categories. Default

> categories are: `commands`, `listeners`, `arguments`, `preconditions`. This example uses the `commands` category.

Reviewer: It's already like that in CLI, you have .sapphirerc.json or .sapphirerc.yml

- Now we add the separator.

```
{
  "category": "commands"
}
---
```

- And the last part, we add the template.

```
{
  "category": "commands"
}
---
import { ApplyOptions } from '@sapphire/decorators';
import { MyExtendedCommand } from './somewhere';
import { Message } from 'discord.js';

@ApplyOptions<MyExtendedCommand.Options>({
	description: 'A basic command'
})
export class {{name}}Command extends MyExtendedCommand {
	public async messageRun(message: Message) {
		return message.channel.send('Hello world!');
	}
}

```

> If you look at the name of the class, you will see it includes `{{name}}`, this is the component's name and it is
> replaced with that name when creating the component. For example: if we created this component with the name
> `HelloWorld`, the name of the exported class would be `HelloWorldCommand`. It is not required but if you need it, this
> is how it's done.
> 
> - And now you can create component with your template

Reviewer 3: ```suggestion
If you look at the name of the class, you will see it includes `{{name}}`. In this context, the formatting variable `name` will be replaced with the command's name when you generate the component. For example, if we created this component with the name `HelloWorld`, the name of the exported class would be `HelloWorldCommand`. Formatting variables are not required, but they're there if you need dynamic templating.

Now you can generate components using your template!
```

> ```
> sapphire generate <component> <name>
> ```

Reviewer 3: ```suggestion
See [the page on generating components][generating-components] for more information.

[generating-components]: ./generating-components.md
```

---

> sidebar_position: 2

Reviewer: `Create custom component templates` should come after `Generating components`
```suggestion
sidebar_position: 3
```

title: Create custom component templates
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem';

> # Create custom component templates
> 

Reviewer: ![](https://reds.is-inside.me/ozVDVhwu.png)
```suggestion
```
